pkg/controller/drain.go

```diff
@@ -700,26 +725,29 @@ func (o *DrainOptions) getVolIDsFromDriver(pvNames []string) ([]string, error) {
 	return o.Driver.GetVolNames(pvSpecs)
 }
 
-func (o *DrainOptions) evictPodInternal(pod *corev1.Pod, policyGroupVersion string, getPodFn func(namespace, name string) (*api.Pod, error), returnCh chan error, stopCh <-chan struct{}) {
+func (o *DrainOptions) evictPodWithoutPVInternal(attemptEvict bool, pod *corev1.Pod, policyGroupVersion string, getPodFn func(namespace, name string) (*api.Pod, error), returnCh chan error) {
 	var err error
 	glog.V(3).Info("Evicting ", pod.Name)
 
-	for {
-		select {
-		case <-stopCh:
-			glog.Warningf("The caller function returned. No need to try evicting pods now")
-			return
-		default:
+	nretries := int(o.MaxEvictRetries)
+	for i := 0; ; i++ {
+		if i >= nretries {
+			attemptEvict = false
+		}
+
+		if attemptEvict {
+			err = o.evictPod(pod, policyGroupVersion)
+		} else {
+			err = o.deletePod(pod)
 		}
-		err = o.evictPod(*pod, policyGroupVersion)
 
 		if err == nil {
 			break
 		} else if apierrors.IsNotFound(err) {
 			glog.V(3).Info("\t", pod.Name, " evicted")
 			returnCh <- nil
 			return
-		} else if apierrors.IsTooManyRequests(err) {
+		} else if attemptEvict && apierrors.IsTooManyRequests(err) {
```

Comment:
Is this the case hit in case of PDB? Just out of curiosity.

Reply:
Yes. This was already existing code. I was suspicious about it that this error is only for rate limiting. But @ggaurav10 found documentation that this is the error that is returned if PD is violated.

Reply:
Okay. Thanks for the clarification.